service/history/execution/mutable_state_util.go

```diff
@@ -437,24 +442,30 @@ func CopyTimerInfo(sourceInfo *persistence.TimerInfo) *persistence.TimerInfo {
 // CopyCancellationInfo copies RequestCancelInfo
 func CopyCancellationInfo(sourceInfo *persistence.RequestCancelInfo) *persistence.RequestCancelInfo {
 	return &persistence.RequestCancelInfo{
-		Version:         sourceInfo.Version,
-		InitiatedID:     sourceInfo.InitiatedID,
-		CancelRequestID: sourceInfo.CancelRequestID,
+		Version:               sourceInfo.Version,
+		InitiatedID:           sourceInfo.InitiatedID,
+		InitiatedEventBatchID: sourceInfo.InitiatedEventBatchID,
+		CancelRequestID:       sourceInfo.CancelRequestID,
 	}
 }
 
 // CopySignalInfo copies SignalInfo
 func CopySignalInfo(sourceInfo *persistence.SignalInfo) *persistence.SignalInfo {
 	result := &persistence.SignalInfo{
-		Version:         sourceInfo.Version,
-		InitiatedID:     sourceInfo.InitiatedID,
-		SignalRequestID: sourceInfo.SignalRequestID,
-		SignalName:      sourceInfo.SignalName,
-	}
-	result.Input = make([]byte, len(sourceInfo.Input))
-	copy(result.Input, sourceInfo.Input)
-	result.Control = make([]byte, len(sourceInfo.Control))
-	copy(result.Control, sourceInfo.Control)
+		Version:               sourceInfo.Version,
+		InitiatedEventBatchID: sourceInfo.InitiatedEventBatchID,
+		InitiatedID:           sourceInfo.InitiatedID,
+		SignalRequestID:       sourceInfo.SignalRequestID,
+		SignalName:            sourceInfo.SignalName,
+	}
+	if sourceInfo.Input != nil {
+		result.Input = make([]byte, len(sourceInfo.Input))
+		copy(result.Input, sourceInfo.Input)
+	}
+	if sourceInfo.Control != nil {
```

Comment:
I checked the `Control` field a few weeks ago to see why/where it is used and couldn't find anything explanatory. Do you know why we carry over this Control field?

Reply:
I have zero idea

Reply:
I tried to figure it out a while ago too :|
I think the answer is somewhere in git history, but I didn't find anything understandable.